File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.ts
function lengthOfLongestSubstring(s: string): number {
    let arrObj = {};
    let cArr: number[] = [];
    let start: number = 0;
    let max: number = 1;
    let len: number = s.length;
    if (!s.length) return 0;
    for (let i = 0; i < len; i++) {
        if (s[i] in arrObj) {
            start = Math.max(start, arrObj[s[i]] + 1);
        } 
        arrObj[s[i]] = i;
        max = Math.max(max, i - start + 1);
        console.log(start, max);
    }
    return max;
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();

        int left = 0;
        int maxLen = 0;

        for (int i = 0; i < s.length(); i++) {
            if (map.containsKey(s.charAt(i)) && map.get(s.charAt(i)) >= left) {
                left = map.get(s.charAt(i)) + 1;
            }

            map.put(s.charAt(i), i);

            maxLen = Math.max(maxLen, i - left + 1);
        }
        return maxLen;
    }
}